commands/fetch_templates.go

```diff
@@ -15,111 +15,154 @@ import (
 	"strings"
 )
 
-const ZipFileName string = "master.zip"
+const (
+	defaultTemplateRepository = "https://github.com/openfaas/faas-cli"
+	templateDirectory         = "./template/"
+)
 
-// fetchTemplates fetch code templates from GitHub master zip file.
-func fetchTemplates(templateUrl string) error {
+var cacheCanWriteLanguage = make(map[string]bool)
 
-	err := fetchMasterZip(templateUrl)
+// fetchTemplates fetch code templates from GitHub master zip file.
+func fetchTemplates(templateURL string, overwrite bool) error {
+	archive, err := fetchMasterZip(templateURL)
+	if err != nil {
+		return err
+	}
 
-	zipFile, err := zip.OpenReader(ZipFileName)
+	zipFile, err := zip.OpenReader(archive)
 	if err != nil {
 		return err
 	}
 
-	log.Printf("Attempting to expand templates from %s\n", ZipFileName)
+	log.Printf("Attempting to expand templates from %s\n", archive)
 
 	for _, z := range zipFile.File {
-		relativePath := strings.Replace(z.Name, "faas-cli-master/", "", -1)
+		var rc io.ReadCloser
+
+		relativePath := z.Name[strings.Index(z.Name, "/")+1:]
+		if strings.Index(relativePath, "template/") != 0 {
+			// Process only directories inside "template" at root
+			continue
+		}
+
+		var language string
+		if languageSplit := strings.Split(relativePath, "/"); len(languageSplit) > 2 {
+			language = languageSplit[1]
+
+			if len(languageSplit) == 3 && relativePath[len(relativePath)-1:] == "/" {
+				// template/language/
+
+				if !canWriteLanguage(language, overwrite) {
+					fmt.Printf("Directory %s exists, overwriting is not allowed\n", relativePath)
+					continue
+				}
+			} else {
+				// template/language/*
+
+				if !canWriteLanguage(language, overwrite) {
+					continue
+				}
+			}
+		} else {
+			// template/
+			continue
+		}
+
 		if strings.Index(relativePath, "template") == 0 {
```

Comment:
Is this `if` statement needed? If `"template"` is not in the `relativePath`, we would have `continue`d  on lines 43-46 above:

```
	relativePath := z.Name[strings.Index(z.Name, "/")+1:]
+		if strings.Index(relativePath, "template/") != 0 {
+			// Process only directories inside "template" at root
+			continue
+		}
```